rfcs/property-getter-setter.md

```diff
@@ -0,0 +1,163 @@
+
+* `get p: T` for a property with getter type `T`, and
+* `set p: T` for a property with setter type `T`.
+
+A property can have both a getter and a setter type. The common case is
+read-write access at the same type:
+
+* `p: T`is the same as `get p: T, set p: T`.
+
+For example:
+```lua
+function f(t)
+  t.r = math.sqrt(t.p)
+  t.p = t.p + t.q
+end
+```
+has inferred type:
+```
+f: (t: { p: number, get q: number, set r: number }) -> ()
+```
+indicating that `p` is used read-write, `q` is used read-only, and `r` is used write-only.
+
+### Indexers
+
+Indexers can be marked `get` or `set` just like properties. In
+particular, this means there are read-only arrays `{get T}`, that are
+covariant, so we have a solution to the "covariant array problem":
+
+```lua
+local dogs: {Dog}
+function f(a: {get Animal}) ... end
+f(dogs)
+```
+
+It is sound to allow this program, since `f` only needs read access to
+the array, and `{Dog}` is a subtype of `{get Dog}`, which is a subtype
+of `{get Animal}`.  This would not be sound if `f` had write access,
+for example `function f(a: {Animal}) a[1] = Cat.new() end`.
+
+### Functions
+
+Functions are not normally mutated after they are initialized, so
+```lua
+function t.f() ... end
+function t:m() ... end
+```
+should introduce a getter, but not a setter. If developers want a mutable function,
+they can use the anonymous function version
+```lua
+t.f = function() ... end
+```
+
+Methods in classes should be read-only by default.
+
+*This is a possibly breaking change.*
+
+### Classes
+
+Getter and setter types apply to classes as well as tables.
+
+Many of the Roblox APIs an be marked as having getters but not
+setters, which will improve accuracy of type checking for Roblox APIs.
+
+### Metatables
+
+Metatables should be read-only by default.
+
+*This is a possibly breaking change.*
+
+### Require
+
+A `require`d module's properties should be read-only by default.
+
+*This is a possibly breaking change.*
+
+### Why separate the getter from the setter?
+
+Separate getters and setters were introduced to TypeScript in response
+to use-cases of APIs which performed data conversion using getter and
+setter methods (the equivalent of Lua `__index` and `__newindex`
+metamethods).
+
+As well as those use-cases, separating read and write access,
+considerably simplifies variance, which is an important part of type inference:
+
+* Read-only properties are *covariant*: if `T` is a subtype of `U`
+  then `{ get p: T }` is a subtype of `{ get p: U }.
+* Write-only properties are *contravariant*: if `T` is a supertype of `U`
+  then `{ set p: T }` is a subtype of `{ set p: U }.
+
+This allows type inference to provide a principal type to tables.
+For example in:
+
+```lua
+local x = { p: nil }
+function g(d : Dog?) x.p = d end
+function h(): Animal? return x.p end
+```
+
+it is not obvious what type to infer for `x`, should it be `{ p: Dog? }` or `{ p: Animal? }`?
```

Comment:
Why is `p: Dog?` not the correct answer?

Reply:
It's not a subtype of `{ p: Animal? }`. There is no principal type here. If we don't have separate getters and setters, then something has to fail to typecheck here: 
```lua
  local t1 : { p: Dog? } = { p = nil }
  local t2 : { p: Animal? } = { p = nil }
  x = t1 -- type error if x : { Animal? }
  x = t2 -- type error if x : { Dog? }
```

Reply:
Maybe I'm missing something. In the example from the RFC, `x.p` can be:
- `nil`, if `g` is never called
- `Dog?`, if `g` is called

As such, it seems correct to say that it stores `Dog?`. `h` typechecks because `Dog?` that's returned from this function is a subtype of `Animal?` which is the declared return type.

Reply:
But if we take `x` to have type `{ p: Dog? }` then the assignment `x = t1` fails to typecheck.

Reply:
Did you mean `x = t2`? If so, yeah, it would fail to typecheck. (note that `t1` and `t2` are absent from the example in RFC)

Reply:
I added them.